Add sorting

sabbaka opened the pull request:
#26 
It's just a first version, will work more on looks.

Here is short demo: https://youtu.be/irg2FrWfOLY

sabbaka commented:
I've added up and down arrow depending on state of sorting. We will need ask Cockpit designers about this also ( how should sortable column title look like ).

spbnick commented:
Thank you, Kyrill, works nicely for me. However, I have two comments/requests:

* The icons looked somewhat confusing to me. I think that using something like triangles pointing up/down (`▲`/`▼`) would be more intuitive, as it would show where the smaller/larger values would be, top or bottom. I wasn't sure what arrows were pointing at.
* When switching between sorting the first or the last column and any other, column sizes change, that shouldn't happen. Column sizes should stay the same regardless of sorting.

sabbaka commented:
1 - done
2 - will need to think a bit.

spbnick commented:
Just to clarify: the column size should stay the same if the displayed data stays the same (if not the order). If the list is paginated and changing sorting brings in the data from another page, then it can change, but not just because sorting icons have changed.

sabbaka commented:
I've fixed changing size issue, please, notice, that is a sort of a hacking of existing Cockpit's component. 
Maybe, there is a sense to add sorting inside this component, yes?
Also, arrow on the last column might look weird, because the last column's text is aligned to the right.

spbnick commented:
Thank you, this looks and feels much nicer! If the code is OK we can merge this for the start.

A couple comments:

* I took a look at the way Kibana sorts its tables and they solved the problem of weird alignment of the last column by simply aligning all of them left. Perhaps we can do that as well.
* If we would still want to have the ability to align columns left or right, then I would say that the data cells can have padding on the right equal to the width of the indicator, so when they align right the column heading would align with data cells, but the indicator would hang over that padding.
* There is no indication in the UI that I can click the columns and change sorting. Looking at Kibana again, perhaps we can show the current sorting when the page is first opened. That would be ascending start timestamp. Then we can show tooltips upon hovering over a heading, saying something like "Sort by Start", or "Sort by End", etc., plus showing the sorting indicator (maybe in light gray), as it would appear if the heading was clicked.
* We would likely need to negotiate this with Cockpit team and perhaps even contribute sorting support to the Cockpit component instead.

Unless you want to implement any of the above right now, please submit those as issues, and I will proceed to code review.

sabbaka commented:
I've added arrow to start column by default and changed text-align for the last column, but I've made it to work with align to the right - now, it's just a matter of personal preference or Cockpit's design practise.

As for reserved space for sorting arrow I'm using div with fixed width - padding won't work, because inner width of element will be changed and padding comes next.

spbnick commented:
Niice, feels much better when stuff doesn't jump around :)

Comments:
* If I come back from a recording to the list, then the sorting indicator (triangle) is not shown. However, since we should probably put sorting setting into the URL anyway, this will be fixed.
* The sorting indicators are back-to-front. Ascending is swapped with descending.

sabbaka commented:
I've fixed arrow issue. Now I'll create issue for URL state for sorting.

spbnick commented on pkg/session_recording/recordings.css:
```diff
@@ -310,3 +310,20 @@ a.disabled {
+    text-decoration: underline;
+}
+
+.sort_icon {
```
Please use dash instead of underscore as in most of the file.

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -340,6 +340,77 @@
+                this.setState({sorting_way: false});
+            } else {
+                this.setState({sorting_way: true});
+            }
```
Just write `this.setState({sorting_way: !this.state.sorting_way})`.

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -340,6 +340,77 @@
+            }
+        }
+
+        sortingBy(field) {
```
Please name functions in imperative mood. I.e. `sortBy` or `setSortingField`, not `sortingBy`.

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -340,6 +340,77 @@
+            this.drawSortArrow = this.drawSortArrow.bind(this);
+            this.state = {
+                sorting_field: "start",
+                sorting_way: "desc",
```
This assignment doesn't match the use below. I assume it should be boolean.

spbnick replied:
The name `sorting_way` is ambiguous, something like `sorting_asc` (`asc` for "ascending") would work better.

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -340,6 +340,77 @@
+            $(this.refs[elem]).append(arrow);
+        }
+
+        sortWay() {
```
Assuming the name of this function starts with a verb in imperative mood (as it should*), it reads like "sort a way", or "sort the way". Judging by its function (ha-ha), it should be named something like `toggleSortingDirection`, or `toggleSortDir`.

> * Another acceptable way to name a function is to start with "is", meaning it is returning a boolean, like "isSortingAsc", but this is not such a function.

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -340,6 +340,77 @@
+            this.sortingBy('duration');
+        }
+
+        sorting(list) {
```
Same here, please use a verb in imperative mood for function name, or start with one. I.e. `sort` or `sortList`.

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -340,6 +340,77 @@
+            let field = this.state.sorting_field;
+            let sort_way = this.state.sorting_way;
+
+            if (sort_way === true) {
```
I know the `sort_way` value can be anything but boolean, but this is taking it a bit too far. I think there is more harm from making such conditions harder to read, than from the possible value type mismatch. So please just write `if (sort_way)` or rather `if (sort_asc)`, and just use `else` for the alternative.

spbnick replied:
If you worry about the value types screwing up, perhaps write a validation function for the whole object and assert the object is valid on function entry and exit. I would leave that as an exercise and a separate PR for later, if you'd like to do that.

Otherwise if we try to handle all the possible value types going sour in place where clear logic should be, we'll get entangled in condition spaghetti.

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -349,10 +420,29 @@
         }
 
+        componentDidUpdate() {
+            this.drawSortArrow(this.state.sorting_field);
```
The `drawSortArrow` function already accesses `this.state.sorting_way`, perhaps we don't need to pass `this.state.sorting_field` to it?

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -349,10 +420,29 @@
+                    className="sort_icon"></div></div>),
+                (<div className="sort" onClick={this.sortByUntil}><span>{_("End")}</span> <div ref="end"
+                    className="sort_icon"></div></div>),
+                (<div className="sort" onClick={this.sortByDuration}><span>{_("Duration")}</span> <div ref="duration"
```
Can we avoid having all these separate handler functions? Perhaps have a single handler and inspect the element being clicked on? Perhaps get the clicked element ID and deduce the column to sort from it?

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -349,10 +420,29 @@
             let rows = [];
+
+            if (this.state.sorting_field != null) {
+                list = list.map(this.calcDuration);
```
I think it is better to just calculate the duration when we load the recording list, in `journalctlIngest`.

sabbaka commented:
Thank you very much for great review! Appreciate that, I've made really some not very efficient decisions.
Now, everything is fixed.

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -459,10 +495,28 @@
             let rows = [];
+
+            if (this.state.sorting_field != null) {
+                this.sortList(list);
```
I think you are not supposed to modify props, so please make a copy of the list above, when assigning to `list`.

spbnick replied:
Or rather just change the `sortList` function into `getSortedList` and let it deal with checking if there's a sorting field, copying, sorting, and returning the (un)sorted list, and then just assign the result to `list`.

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -602,6 +656,7 @@
                          /* FIXME Should be start + message duration */
                          end:       ts};
                     /* Map the recording */
+                    r.duration = r.end - r.start;
```
This would be zero upon creation, so you can just put `duration: 0` above.

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -450,6 +450,42 @@
+            $(this.refs[this.state.sorting_field]).append(arrow);
+        }
+
+        sortBy(event) {
```
This is no longer just a `sortBy` function. First of all, it now takes a non-trivial object, a click event as the argument, then extracts the sorting field from it, and it also changes the sorting order. So, I think it should be called something more specific, like `handleColumnClick`.

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -450,6 +450,42 @@
+
+        sortList(list) {
+            let field = this.state.sorting_field;
+            let sort_way = this.state.sorting_asc;
```
The function deals with sorting exclusively, so we don't need to have `sort_` in the variable name, but we need to change from `way` to `asc`. I.e. I think it would be better to call this variable `asc` instead.

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -450,6 +450,42 @@
+            };
+        }
+
+        drawSortArrow() {
```
Since the indicator is no longer, strictly speaking, an arrow, perhaps we should rename this to `drawSortDir`, `drawSortIndicator`, or some such.

sabbaka commented:
Fixed, please, take a look.

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -450,6 +450,47 @@
+            this.drawSortDir = this.drawSortDir.bind(this);
+            this.state = {
+                sorting_field: "start",
+                sorting_asc: false,
```
This is opposite to the order `this.props.list` is in, did you intend it to be so?

spbnick commented on pkg/session_recording/recordings.jsx:
```diff
@@ -450,6 +450,47 @@
+        handleColumnClick(event) {
+            this.setState({
+                sorting_field: event.currentTarget.id,
+                sorting_asc: !this.state.sorting_asc
```
I think this is behaving somewhat confusingly. If I click on an unsorted column, I would most likely expect it to be sorted in a particular way, and not in a way opposite from the previously sorted column. It might be dependent on a column type, but perhaps easier to understand and remember can be just ascending order.

sabbaka commented:
I've made requested changes.

spbnick commented:
Thank you, Kyrill, works and looks great now, merged :)!
